### PR TITLE
Remove -unittest special cases wrt. template emission

### PR DESCRIPTION
**Pull request by kinke**
One part of #11362.

### CODE REVIEW

**Comment by atilaneves** [comment]
Rebase?

**Comment by kinke** [comment]
### Rationale

The template emission with `-unittest` compiles was very conservative, disabling template codegen culling in many cases. This is particularly costly for separate per-module `-unittest` compilations. The likely motivation is that an existing instantiation isn't guaranteed to have been compiled with `-unittest` too, so `version (unittest)` blocks and `unittest` functions *inside the template* might not have been instantiated, so the compiler goes the safe route if in doubt.

The idea here is to simply ignore this 'danger' by considering `version (unittest)` blocks and `unittest` *inside templates* as code-smell (without deprecations/warnings for now ;)), and to remove all related template-emission special cases for `-unittest`.  E.g., `Nullable` unittests *inside* the template (https://github.com/dlang/phobos/blob/1c1ab745387e07acd6de135516da6bf93c1fbbe5/std/typecons.d#L2840-L2892) are no longer pointlessly instantiated (pointlessly because not depending on template param `T`) and run for each `Nullable` instantiation (easily hundreds) in user code being unittest-compiled. Instead, these unittests are only respected when compiling the `std.typecons` module with `-unittest`, and run if there's at least one emitted `Nullable` instantiation in that compilation batch.

This ignoring of unittests in non-root modules is required, otherwise if the compiler determines that a `Nullable` instantiation (semantically analyzed with `-unittest`) can be culled due to an existing instantiation, it would also consider all other template instantiations *in the unittests* as candidates for further culling, leading to dreaded missing symbols at link-time.

**Comment by kinke** [comment]
I've ported this to LDC (https://github.com/ldc-developers/ldc/pull/3860) and gathered some numbers for the change in emitted code. I've compiled druntime and Phobos, each module separately, and then compared the aggregated size of all object files. Findings:

* With `-unittest`: about -8.7% (almost identical percentage across release and debug versions) with the new template emission alone, -11% with additional ignoring of non-root unittests. New `-allinst` adds some 0.1% to the new size and thus seems to hardly make a difference.
* Without `-unittest`: +0.5% with the new template emission, and new `-allinst` again hardly changes that number.

**Comment by atilaneves** [comment]
I wrote a script to try out these changes on several projects that are on the buildkite list. The results are below. I didn't bother to exclude link times since it would be hard to do for (IMHO) little benefit. "Build" was done with `dub build` and "test" with `dub build --build=unittest`, except for my own projects where `-c unittest` was added (I couldn't add it to all projects because some don't have an explicit unittest configuration). Results for memory and build times were taken from the minimum for both after 3 builds. 

All builds were "all at once" and from scratch (after `git clean -xffd`). Incremental per-module or per-package builds would see more drastic improvements. I think that given these results this PR is a no-brainer.

```
dlang/dub:
    build          0.00% mem     -0.64% time
    test          -5.44% mem     -2.53% time
dlang-community/DCD:
    build          0.89% mem      0.68% time
    test         -16.03% mem     -6.94% time
dlang-community/libdparse:
    build         -0.73% mem     -1.53% time
    test          -8.10% mem     -6.27% time
dlang-community/dfmt:
    build          0.21% mem     -0.50% time
    test          -7.06% mem     -4.29% time
atilaneves/automem:
    build          0.03% mem     -2.38% time
    test         -16.41% mem    -13.38% time
atilaneves/unit-threaded:
    build         -0.79% mem     -1.04% time
    test         -23.87% mem    -12.83% time
rejectedsoftware/ddox:
    build         -0.01% mem      0.32% time
    test         -21.91% mem    -10.25% time
rejectedsoftware/diet-ng:
    build         -0.78% mem     -3.08% time
    test         -23.69% mem    -18.77% time
BlackEdder/ggplotd:
    build         -1.93% mem      1.08% time
    test         -12.56% mem    -10.02% time
eBay/tsv-utils:
    build         -0.08% mem      0.90% time
    test         -36.11% mem    -22.97% time
d-widget-toolkit/dwt:
    build         -0.07% mem      0.63% time
    test          -7.22% mem     -5.39% time
mbierlee/poodinis:
    build         -0.48% mem      0.00% time
    test         -35.64% mem    -26.55% time
gecko0307/dagon:
    build         -0.07% mem     -0.77% time
    test         -17.32% mem    -15.88% time
jmdavis/dxml:
    build         -0.71% mem     -2.27% time
    test          -4.92% mem     -3.29% time
libmir/mir-algorithm:
    build         -0.27% mem     -1.55% time
    test          -1.43% mem     -1.69% time
libmir/mir-random:
    build         -0.12% mem      1.85% time
    test         -12.47% mem     -9.90% time
libmir/mir-core:
    build          0.04% mem     -7.32% time
    test          -7.73% mem     -8.28% time
Abscissa/libInputVisitor:
    build          0.13% mem     -2.78% time
    test           0.01% mem     -7.89% time
PhilippeSigaud/Pegged:
    build         -0.38% mem     -1.05% time
    test          -3.50% mem     -0.04% time
```

**Review by RazvanN7 — APPROVED** [review]
This is awesome work! Thanks!

**Review by WalterBright — DISMISSED** [review]
@kinke I'll approve if your rationale is added as a comment somewhere in the code. Thanks!

File: src/dmd/dtemplate.d [hunk 1]
```diff
@@ -6264,73 +6268,33 @@ extern (C++) class TemplateInstance : ScopeDsymbol
 
         if (global.params.allInst)
         {
-            return true;
```
Comment (MoonlightSentinel) [inline src/dmd/dtemplate.d:6267]: Why was this changed - this isn't related to the unittest hack? IIUC the changes correctly `-allinst` will no longer be a valid workaround for missing template instances caused by instance expected from other (read as non-root) modules.

**Review by MoonlightSentinel — DISMISSED** [review]
The description in `cli.d` must be updated to reflect the changes to `-allinst` (if the proposed change is desired).

File: src/dmd/dtemplate.d (hunk 1 above)
Comment (kinke) [reply src/dmd/dtemplate.d:6267]: > IIUC the changes correctly -allinst will no longer be a valid workaround

Atila asked me to look into the templates emission stuff because `-allinst` made things worse for one of his earlier experiments - more missing symbols, not less. So the assumption is that it now works as people would expect it to. Before rebasing, I've let the CI testsuite run with an enforced `-allinst`, and it was all green.

Comment (kinke) [reply src/dmd/dtemplate.d:6267]: Elaborating a bit: after removing the `-unittest` special cases for template emission, all hell broke loose, lots of undefined symbols. It took a while to tweak the default (non-`allinst`) template emission algorithm to get rid of most of them; ignoring non-root unittests fixed the last CI failures. My results for druntime/Phobos (+0.5% code size without `-unittest`) show that I had to cull a bit less to get there.

So I hope the situation wrt. spurious missing undefined template symbols (especially for separate per-module compilations) improves with these changes. Coverage is definitely better, because all `-unittest` compiles during CI now test the default template emission algorithm.

---

`-allinst` was supposed to work around these issues, but might make things worse - not just for initial experiments here, but also for https://issues.dlang.org/show_bug.cgi?id=19123, https://issues.dlang.org/show_bug.cgi?id=19406, …

The algorithm I now use for `-allinst` is what I expected it to be - emitting *all* templates instantiated in some root module. So when e.g. compiling 10 modules (to 1 or 10 object files), the set of object files should be self-sufficient wrt. instantiated symbols - not referencing any external templated symbol at all.

Comment (MoonlightSentinel) [reply src/dmd/dtemplate.d:6267]: Thanks for the detailed writeup.

Initially thought about the problem of linker errors introduced by `-checkaction=context`, i.e. templates expected from druntime / phobos which don't include them due to `-release` or `-checkaction=D` (e.g. for LDC with `--link-defaultlib-debug`). `-allinst` used to be a viable workaround but now has no effect - the templates are missing  (even when combined with  `-i` / explicitly passing the module) for `master` *and*  also this patch.

Seems like this patch doesn't make it worse for `-checkaction=context` and improves for everything else.

**Comment by RazvanN7** [comment]
@MoonlightSentinel Are we good to go?

File: src/dmd/dtemplate.d (hunk 1 above)
Comment (kinke) [reply src/dmd/dtemplate.d:6267]: > the templates are missing (even when combined with -i / explicitly passing the module) for master and also this patch

Can you post reproduction steps or file an issue? Thanks!

Comment (MoonlightSentinel) [reply src/dmd/dtemplate.d:6267]: Use an `assert` with the same types as done in any module imported from Druntime / Phobos., e.g. `assert(someInt)`

See [19937](https://issues.dlang.org/show_bug.cgi?id=19937) and [22374](https://issues.dlang.org/show_bug.cgi?id=22374).

**Comment by MoonlightSentinel** [comment]
I think the documentation of `-allinst` needs to be updated but that also can be deferred to another PR.

**Comment by maxhaton** [comment]
I think there is a bug report for this in bugzilla somewhere.

File: src/dmd/dtemplate.d (hunk 1 above)
Comment (kinke) [reply src/dmd/dtemplate.d:6267]: Thx, can reproduce with issue 22374 and `-allinst` - with this patch, 2 undefined template symbols instantiated inside `_d_assert_fail` (`inFinalizer!()` and some `miniFormatFakeAttributes`). When modding `needsCodegen()` to `return true` in all cases, the same 2 symbols remain undefined (and many more). So the problem for these lie somewhere else; apparently they are not appended to the root module's members and so no codegen candidates.

**Comment by ibuclaw** [comment]
This seems to have broken the std.datetime.date unittest - and only that module - in gdc's testsuite, and I'm not immediately sure why.

**Comment by MoonlightSentinel** [comment]
https://issues.dlang.org/show_bug.cgi?id=22730

**Comment by ibuclaw** [comment]
This PR caused a regression: https://issues.dlang.org/show_bug.cgi?id=23255

**Comment by ibuclaw** [comment]
The PR broke the "workaround" in the following issue: https://issues.dlang.org/show_bug.cgi?id=15985
